cluster_tools/schedulers/slurm.py

```diff
@@ -100,22 +160,61 @@ def inner_submit(
             for resource, value in self.job_resources.items():
                 job_resources_lines += ["#SBATCH --{}={}".format(resource, value)]
 
-        job_array_line = ""
-        if job_count is not None:
-            job_array_line = "#SBATCH --array=0-{}".format(job_count - 1)
-
-        script_lines = (
-            [
-                "#!/bin/sh",
-                "#SBATCH --output={}".format(log_path),
-                '#SBATCH --job-name "{}"'.format(job_name),
-                job_array_line,
-            ]
-            + job_resources_lines
-            + [*additional_setup_lines, "srun {}".format(cmdline)]
+        max_array_size = self.get_max_array_size()
+        max_submit_jobs = self.get_max_submit_jobs()
+        # Only ever submit at most a third of max_submit_jobs at once.
+        # This way, multiple programs submitting slurm jobs will not block each other
+        # by "occupying" more than half of the number of submittable jobs.
+        batch_size = min(max_array_size, max_submit_jobs // 3)
+
+        scripts = []
+        job_id_futures = []
+        ranges = []
+        number_of_jobs = job_count if job_count is not None else 1
+        for job_index_start in range(0, number_of_jobs, batch_size):
+            # job_index_end is inclusive
+            job_index_end = min(job_index_start + batch_size, number_of_jobs) - 1
+            array_index_end = job_index_end - job_index_start
+
+            job_array_line = ""
+            if job_count is not None:
+                job_array_line = "#SBATCH --array=0-{}".format(array_index_end)
+
+            script_lines = (
+                [
+                    "#!/bin/sh",
+                    "#SBATCH --output={}".format(log_path),
+                    '#SBATCH --job-name "{}"'.format(job_name),
+                    job_array_line,
+                ]
+                + job_resources_lines
+                + [
+                    *additional_setup_lines,
+                    "srun {} {}".format(cmdline, job_index_start),
+                ]
+            )
+
+            job_id_futures.append(futures.Future())
+            scripts.append("\n".join(script_lines))
+            ranges.append((job_index_start, job_index_end + 1))
+
+        job_sizes = [end - start for start, end in ranges]
+        # Replace with asyncio.run when upgrading to python 3.7+
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(
+            self.submit_once_allowed(scripts, job_sizes, job_id_futures)
         )
 
-        return self.submit_text("\n".join(script_lines))
+        return job_id_futures, ranges
+
+    async def submit_once_allowed(self, scripts, job_sizes, futures):
```

Comment:
cool :+1: probably, the first time we use asyncio, but seems like a good fit here :)

Reply:
hm, on a second thought: won't `inner_submit` block until all job ids are available due to `run_until_complete`? then, I don't really see the point of returning futures. also, client code could only react to job results after all jobs were scheduled. I think, the code should submit the futures into the event loop without blocking. Maybe `asyncio.run_coroutine_threadsafe` (https://docs.python.org/3.5/library/asyncio-task.html#asyncio.run_coroutine_threadsafe) fits here?